src/vuln_analysis/utils/java_utils.py

```diff
@@ -0,0 +1,2189 @@
+import hashlib
+import re
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Optional, Any
+from langchain_core.documents import Document
+from tqdm import tqdm
+
+PRIMITIVE_TYPES = {"byte","short","int","long","float","double","boolean","char","void"}
+
+# Accepts:
+# - groupId: dot-separated segments; each segment alnum, may contain '-' or '_' inside
+# - artifactId: alnum, may contain '.', '_' or '-' inside
+# - version: alnum start, may contain '.', '_' or '-' (e.g., 1.2.3, 1.0.0-SNAPSHOT)
+_GAV_RE = re.compile(
+    r"""
+        ^(?P<group>
+            (?:[A-Za-z0-9]+(?:[A-Za-z0-9_-]*[A-Za-z0-9])?)
+            (?:\.(?:[A-Za-z0-9]+(?:[A-Za-z0-9_-]*[A-Za-z0-9])?))*
+        )
+        :
+        (?P<artifact>[A-Za-z0-9]+(?:[A-Za-z0-9._-]*[A-Za-z0-9])?)
+        :
+        (?P<version>[A-Za-z0-9][A-Za-z0-9._-]*)
+        $
+        """,
+    re.X,
+)
+
+@dataclass
+class ParseResult:
+    kind: str                  # "class" | "interface" | "enum" | "record" | "method" | "unknown"
+    name: Optional[str] = None # simple name if found (class or method)
+    meta: Optional[Dict[str, Any]] = None
+
+def add_missing_jar_string(s: str) -> str:
+    i = s.rfind(':')
+    if i == -1:
+        return s
+    return s[:i] + ':jar' + s[i:]
+
+
+def convert_from_maven_artifact(package_name: str) -> str:
+    s = package_name
+
+    # 1) Remove substring up to and including the first ':'
+    first = s.find(":")
+    if first != -1:
+        s = s[first + 1:]
+
+    # 2) Replace the last ':' (if any) with '-'
+    last = s.rfind(":")
+    if last != -1:
+        s = s[:last] + "-" + s[last + 1:]
+
+    return s
+
+def convert_to_maven_artifact(package_name: str) -> str:
```

Comment:
@tmihalac This is only used internally here, please refactor to start with underscore for a better readability and to reflect intention.
If you think that you're going to use it outside of this module in the future, you can leave it as is...

Reply:
OK